File: server/api/pkg/diagnostic/diagnostic.go
package diagnostic

import (
	"time"

	"github.com/reearth/reearth-flow/api/pkg/id"
)

type SourceSpan struct {
	length *uint
	offset uint
}

func NewSourceSpan(offset uint, length *uint) *SourceSpan {
	return &SourceSpan{offset: offset, length: length}
}

func (s *SourceSpan) Offset() uint {
	return s.offset
}

func (s *SourceSpan) Length() *uint {
	return s.length
}

type AggregateInfo struct {
	sampleFeatureIDs []string
	count            uint64
}

func NewAggregateInfo(count uint64, sampleFeatureIDs []string) *AggregateInfo {
	return &AggregateInfo{count: count, sampleFeatureIDs: sampleFeatureIDs}
}

func (a *AggregateInfo) Count() uint64 {
	return a.count
}

func (a *AggregateInfo) SampleFeatureIDs() []string {
	return a.sampleFeatureIDs
}

// NodeID/Category/Severity/EffectiveDisposition are plain strings, not
// id.NodeID or enums: engine node IDs aren't always UUIDs, and unknown
// engine values must round-trip verbatim.
type Diagnostic struct {
	timestamp            time.Time
	featureID            *string
	sourceSpan           *SourceSpan
	effectiveDisposition *string
	nodeID               *string
	actionType           *string
	aggregated           *AggregateInfo
	help                 *string
	code                 string
	category             string
	severity             string
	message              string
	jobID                id.JobID
	// terminal is internal-only, never wire/GraphQL-exposed.
	terminal bool
}

func (d *Diagnostic) JobID() id.JobID {
	return d.jobID
}

func (d *Diagnostic) NodeID() *string {
	return d.nodeID
}

func (d *Diagnostic) Timestamp() time.Time {
	return d.timestamp
}

func (d *Diagnostic) Code() string {
	return d.code
}

func (d *Diagnostic) Category() string {
	return d.category
}

func (d *Diagnostic) Severity() string {
	return d.severity
}

func (d *Diagnostic) EffectiveDisposition() *string {
	return d.effectiveDisposition
}

func (d *Diagnostic) ActionType() *string {
	return d.actionType
}

func (d *Diagnostic) FeatureID() *string {
	return d.featureID
}

func (d *Diagnostic) Message() string {
	return d.message
}

func (d *Diagnostic) Help() *string {
	return d.help
}

func (d *Diagnostic) Aggregated() *AggregateInfo {
	return d.aggregated
}

func (d *Diagnostic) SourceSpan() *SourceSpan {
	return d.sourceSpan
}

func (d *Diagnostic) Terminal() bool {
	return d.terminal
}

File: ui/src/types/diagnostic.ts
/**
 * A structured diagnostic emitted by the engine: a per-feature error or
 * warning, a `finish()`-time aggregated summary, or a terminal per-node
 * failure.
 *
 * `category`, `severity` and `effectiveDisposition` are plain strings on the
 * wire rather than GraphQL enums, deliberately: a newer engine can emit a
 * value this build has never heard of and it still has to round trip. The cost
 * is that there is no compile-time exhaustiveness check on them, so every
 * consumer below recognises the values we know about and passes anything else
 * through verbatim.
 */
export type Diagnostic = {
  code: string;
  category: string;
  /**
   * Display level only. Never branch on this to decide whether a run passed or
   * failed — use {@link isFatalDiagnostic}, which reads
   * `effectiveDisposition`.
   */
  severity: string;
  /** The authoritative fatality signal. Unresolved (`undefined`) for warn-and-continue. */
  effectiveDisposition?: string;
  nodeId?: string;
  actionType?: string;
  featureId?: string;
  message: string;
  help?: string;
  /**
   * How many times this row's error code fired, on the rows that know: a
   * `finish()`-time roll-up (e.g. "1,204 features dropped, 5 samples") and a
   * node's terminal failure, which counts the features that failed with its
   * code. Absent on a per-feature row reported on its own. Read it structurally
   * — never parse the count out of `message`.
   */
  aggregatedCount?: number;
  sampleFeatureIds?: string[];
};

/**
 * Whether this diagnostic is what failed the run. This is the only correct
 * pass/fail test: a diagnostic can carry `severity: "fatal"` for display while
 * a node-level policy downgraded its disposition, and vice versa.
 */
export const isFatalDiagnostic = (diagnostic: Diagnostic): boolean =>
  diagnostic.effectiveDisposition === "fatal";

/** Whether this row stands for a counted set of occurrences rather than one unquantified event. */
export const isAggregatedDiagnostic = (diagnostic: Diagnostic): boolean =>
  diagnostic.aggregatedCount !== undefined;

/**
 * How many occurrences a row stands for, or `undefined` when the row carries no
 * count.
 *
 * Never substitute 1 for `undefined`. A row without a count is not a row that
 * happened once — it is a row whose payload does not say, and rendering it as 1
 * states something the engine did not. Show it as unknown instead.
 */
export const diagnosticOccurrences = (
  diagnostic: Diagnostic,
): number | undefined => diagnostic.aggregatedCount;

/**
 * Rank used for ordering and for picking the worst diagnostic in a set.
 * Unknown severities sort below every known one rather than being dropped.
 */
const severityRank: Record<string, number> = {
  trace: 0,
  debug: 1,
  info: 2,
  warn: 3,
  error: 4,
  fatal: 5,
};

export const diagnosticSeverityRank = (severity: string): number =>
  severityRank[severity] ?? -1;

/**
 * Whether a severity should read as a failure rather than an advisory. Purely a
 * display decision — pass/fail is {@link isFatalDiagnostic}, which reads
 * `effectiveDisposition`. Shared so a badge and a node marker cannot disagree
 * about what counts as bad.
 */
export const isBlockingSeverity = (severity?: string): boolean =>
  severity === "fatal" || severity === "error";

/** Sorts worst-first, so the row a user needs to read is at the top. */
export const compareDiagnosticSeverity = (
  a: Diagnostic,
  b: Diagnostic,
): number =>
  diagnosticSeverityRank(b.severity) - diagnosticSeverityRank(a.severity);
